honeybadgermpc/offline.py

```diff
@@ -0,0 +1,176 @@
+        assert type(inputs) in [tuple, list]
+        avss_tasks = []
+        avss_tasks.append(asyncio.create_task(
+            self.avss_instance.avss_parallel(
+                avss_id, len(inputs), values=inputs, dealer_id=self.my_id)))
+        for i in range(self.n):
+            if i != self.my_id:
+                avss_tasks.append(asyncio.create_task(
+                    self.avss_instance.avss_parallel(
+                        avss_id, len(inputs), dealer_id=i)))
+        await asyncio.gather(*avss_tasks)
+        logging.debug("[%d] AVSS for offline phase completed - %d", self.my_id, avss_id)
+
+    async def _runner(self):
+        counter = 0
+        logging.debug("[%d] Starting preprocessing runner: %s", self.my_id, self.tag)
+        while True:
+            # If the number of values in the output queue are below the lower
+            # watermark then we want to trigger the next set of AVSSes.
+            if self.output_queue.qsize() < self.low_watermark:
+                logging.debug("[%d] Starting AVSS: %d", self.my_id, counter)
+                await self._trigger_and_wait_for_avss(counter)
+                logging.debug("[%d] AVSS Completed: %d", self.my_id, counter)
+                counter += 1
+            # Wait for sometime before checking again.
+            await asyncio.sleep(PreProcessingBase.PERIOD_IN_SECONDS)
+
+    async def _get_output_batch(self, group_size=1):
+        for i in range(self.batch_size):
+            batch = []
+            while True:
+                value = await self.avss_value_processor.get()
+                if value is None:
+                    break
+                batch.append(value)
+            assert len(batch) / group_size >= self.n - self.t
+            assert len(batch) / group_size <= self.n
+            yield batch
+
+    async def _extract(self):
+        raise NotImplementedError
+
+    def __enter__(self):
+        n, t, my_id = self.n, self.t, self.my_id
+        send, recv = self.get_send_recv(f'{self.tag}-AVSS')
+        g, h, pks, sk = get_avss_params(n, t, my_id)
+        self.avss_instance = HbAvssLight(pks, sk, g, h, n, t, my_id, send, recv)
+        self.avss_instance.__enter__()
+        self.tasks.append(asyncio.create_task(self._runner()))
+
+        send, recv = self.get_send_recv(f'{self.tag}-AVSS_VALUE_PROCESSOR')
+        pk, sks = dealer(n, t+1, seed=17)
+        self.avss_value_processor = AvssValueProcessor(
+            pk, sks[my_id],
+            n, t, my_id,
+            send, recv,
+            self.avss_instance.output_queue.get,
+            self.avss_value_processor_chunk_size)
+        self.avss_value_processor.__enter__()
+        self.tasks.append(asyncio.create_task(self._extract()))
+        return self
+
+    def __exit__(self, *args):
+        for task in self.tasks:
+            task.cancel()
+        self.avss_instance.__exit__(*args)
+        self.avss_value_processor.__exit__(*args)
+
+
+class RandomGenerator(PreProcessingBase):
```

Comment:
This is particularly elegant - it exactly matches the collect inputs -> extract randomness description from the papers